syntax = "proto3";

import "google/protobuf/struct.proto";
import "google/protobuf/any.proto";

package simulation;

service SimulationService {
    rpc StreamSimulation (SimulationRequest) returns (stream SimulationUpdate);
}

message Object {
    // repeated string path = 2;
    google.protobuf.Any value = 1;
}

message Process {
    string type = 1;
    string address = 2;
    map<string, google.protobuf.Struct> config = 3;
    map<string, Object> inputs = 4;
    map<string, Object> outputs = 5;
}

message Result {
    map<string, google.protobuf.Struct> data = 1;
}

message Results {
    repeated Result data = 1;
}

message SimulationRequest {
    string job_id = 1;
    string last_updated = 2;
    repeated string simulators = 3;
    int32 duration = 4;
    map<string, Process> spec = 5;
}

message SimulationUpdate {
    string job_id = 1;
    string last_updated = 2;
    Results results = 3;
}
